python/GafferUI/NumericWidget.py

```diff
@@ -304,3 +312,70 @@ def __emitValueChanged( self, reason ) :
 			return
 
 		signal( self, reason )
+
+# A basic validator/evaluator that supports simple maths
+# operators [+-/*] along with standard number validation, eg:
+#   2 + 3
+#   4.4 / 2
+class BasicNumericMathExpression( QtGui.QValidator ) :
+
+	__invalid = re.compile( r"[^0-9\.+\-*/\s]" )
+	__operators = re.compile( r"[+\-/*]" )
+	__whitespace = re.compile( r"\s+" )
+
+	def __init__( self, parent, typeValidator ) :
+
+		QtGui.QValidator.__init__( self, parent )
+
+		self.__typeValidator = typeValidator
+		self.__typeValidator.setParent( self )
+
+	def validate( self, text, pos ) :
+
+		if not self.__isExpression( text ) :
+			return self.__typeValidator.validate( text, pos )
+
+		# Invalid chars
+		if re.search( self.__invalid, text ) is not None :
+			return QtGui.QValidator.Invalid, text, pos
+
+		# Remove whitespace to simplify tokenisation
+		text = re.sub( self.__whitespace, "", text )
+
+		# Ends with an operator
+		if re.match( self.__operators, text[ -1 ] ) :
+			return QtGui.QValidator.Intermediate, text, pos
+
+		# Individual numbers invalid for our type validator
+		for number in re.split( self.__operators, text ) :
+			# May be seen with *-1, etc...
+			if number == '' :
+				continue
+			state = self.__typeValidator.validate( number, len(number)-1 )
+			if state[0] != QtGui.QValidator.Acceptable :
+				return state[0], text, pos
+
+		# Some other syntax error
+		try :
+			self.eval( text )
+		except SyntaxError :
+			return QtGui.QValidator.Invalid, text, pos
+
+		return QtGui.QValidator.Acceptable, text, pos
+
+	def eval( self, text ) :
+
+		if not self.__isExpression( text ) :
+			return text
```

Comment:
This seems confusing - why sometimes return a string and sometimes a number?

Reply:
Conformed [here](https://github.com/GafferHQ/gaffer/pull/3873/commits/2b3a1aebbc22fb20c09867ffa5f982595388a119).